### PR TITLE
[Enhance] Speedup the Video Inference by Accelerating data-loading Stage in Top_Down_Full_Frame_without_Det

### PR DESCRIPTION
**Pull request by chenxinfeng4**
It's a parallel to my mmdetection PR.
## Motivation

The video inference was not efficient, because there are many "resize", "padding", "to rgb" and "normalize" operations in CPU workload. Besides, the 'img_metas' is calculated for every frame, which is nonefficient. The CPU workload is really high, but the speed is slow.

## Modification

I transformed the video "crop", "resize", "padding", "to rgb" in ffmpeg-based video reader, which is lite and CPU friendly. The video reader can support NVIDIA-VIDEO-DECODING if possible, to save more CPU workload. And the "normalize" is done in GPU. In all, I reduced the data-loading time and CPU workload.

## Result

For the pure data loading, it can reach 400 Fps (possible the ceiling of NVIDIA-DECODER). The CPU work load is acceptable (450%).
```python
    for frame_resize, frame_origin in zip(
            mmcv.track_iter_progress(video_resize), video_origin):

        # test a single frame without model inference
        data = process_img(frame_resize, img_metas, args.device)
        continue
        pose_results = pose_model(
                return_loss=False, return_heatmap=False, **data)
```
> The video size is 1280x800, converting to image size 512x512. 

图片

图片

## Limitation
1. Only works for `top_down_video_demo_full_frame_without_det.py`. May be adaptive to `bottom_up_video_demo.py`. 
2. Didn't support for flip-pair test.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.

**After PR**:

- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by chenxinfeng4** [comment]
The whole inference speed is 20Fps in my version, and 14Fps in original CPU data loading version. By combining other tricks, e.g. batchsize=8, fp16 (not post in this PR), it can run at 120Fps.

 This PR is similar to my previous PR #1320 ：
本次PR和 #1320 都是优化inference阶段数据加载。1. 本次PR的将所有改动全部整合到一个文件，减少对其它代码干扰；2. 本次PR主要解决 video resize/padding 的加速；3. 因为 network inference 速度瓶颈存在，本次PR和 #1320 没有体验的差异；4.  network inference 经过Tensorrt加速后最高可到300Fps，此时本次PR明显优于  #1320 。

File: demo/top_down_video_demo_full_frame_without_det_gpuaccel.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    frame_cuda = F.normalize(frame_cuda, mean=mean, std=std, inplace=True)
+    frame_cuda = frame_cuda[None, :, :, :]  # NCHW
+    data = {'img': frame_cuda, 'img_metas': [img_metas]}
+    return data
+
+
+def main():
+    """Visualize the demo video with GPU acceleration.
+
+    Using mmdet to detect the human.
```
Comment (liqikai9) [inline demo/top_down_video_demo_full_frame_without_det_gpuaccel.py:88]: Seems like we don't need `mmdet` here to detect the human.

Comment (chenxinfeng4) [reply demo/top_down_video_demo_full_frame_without_det_gpuaccel.py:88]: I copy that from original `top_down_video_demo_full_frame_without_det.py`

File: demo/top_down_video_demo_full_frame_without_det_gpuaccel.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    pose_model.cfg.data['test']['dataset_info'] = dataset_info
+    if args.nvdecode:
+        VideoCapture = ffmpegcv.VideoCaptureNV
+    else:
+        VideoCapture = ffmpegcv.VideoCapture
+    video_origin = VideoCapture(args.video_path)
+    img_metas = prefetch_img_metas(pose_model.cfg,
+                                   (video_origin.width, video_origin.height))
+    resize_wh = pose_model.cfg.data_cfg['image_size']
+    video_resize = VideoCapture(
```
Comment (liqikai9) [inline demo/top_down_video_demo_full_frame_without_det_gpuaccel.py:147]: Is this `resize` operation in video reading equivalent to the `TopDownAffine` pipeline in MMPose?

File: demo/top_down_video_demo_full_frame_without_det_gpuaccel.py (hunk 1 above)
Comment (liqikai9) [reply demo/top_down_video_demo_full_frame_without_det_gpuaccel.py:88]: OK, I see. Have you tested this demo script with some config? Does the visualization result look fine?

Comment (chenxinfeng4) [reply demo/top_down_video_demo_full_frame_without_det_gpuaccel.py:88]: This code has the same API with the original `top_down_video_demo_full_frame_without_det.py`. I've tested this code with the some config of mine and got the identical results. I hope this code will be general to other configs.

File: demo/top_down_video_demo_full_frame_without_det_gpuaccel.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+except ImportError:
+    raise ImportError(
+        'Please install ffmpegcv with:\n\n    pip install ffmpegcv')
+
+
+def prefetch_img_metas(cfg, ori_wh):
+    w, h = ori_wh
+    bbox = np.array([0, 0, w, h])
+    center, scale = _box2cs(cfg, bbox)
+    scale /= 1.25  # never expand bbox area
```
[the remark this answers is not in the document]
Comment (chenxinfeng4) [reply demo/top_down_video_demo_full_frame_without_det_gpuaccel.py:27]: The original ` _box2cs` method will give a 1.25x bigger box than original one. Which means add black borders for each frame in every directions. By `scale /= 1.25 ` , I align the `scale` to match the rest of my code. I've checked that myself.
```python
# mmpose/apis/inference.py
def _box2cs(cfg, box):
    x, y, w, h = box[:4]
    input_size = cfg.data_cfg['image_size']
    aspect_ratio = input_size[0] / input_size[1]
    center = np.array([x + w * 0.5, y + h * 0.5], dtype=np.float32)

    if w > aspect_ratio * h:
        h = w * 1.0 / aspect_ratio
    elif w < aspect_ratio * h:
        w = h * aspect_ratio

    # pixel std is 200.0
    scale = np.array([w / 200.0, h / 200.0], dtype=np.float32)
    scale = scale * 1.25   # <====HERE====

    return center, scale
```

[the remark this answers is not in the document]
Comment (chenxinfeng4) [reply demo/top_down_video_demo_full_frame_without_det_gpuaccel.py:27]: However, I see the `_box2cs` may has been changed in recent PR, and the `1.25` can be set in config file.

File: demo/top_down_video_demo_full_frame_without_det_gpuaccel.py (hunk 2 above)
Comment (chenxinfeng4) [reply demo/top_down_video_demo_full_frame_without_det_gpuaccel.py:147]: Yes, the `resize` is a short cut version of `TopDownAffine`. In the video inference, most people don't need 'flip', 'rotation' operations. In this case, **the `TopDownAffine` can be degenerated as video resize and padding**, which is implemented by high performance ffmpeg now.

File: demo/top_down_video_demo_full_frame_without_det_gpuaccel.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+from mmpose.datasets import DatasetInfo
+
+try:
+    import ffmpegcv
+except ImportError:
+    raise ImportError(
+        'Please install ffmpegcv with:\n\n    pip install ffmpegcv')
+
+
+def prefetch_img_metas(cfg, ori_wh):
```
[the remark this answers is not in the document]
Comment (chenxinfeng4) [reply demo/top_down_video_demo_full_frame_without_det_gpuaccel.py:23]: OK, I will add that.

File: demo/top_down_video_demo_full_frame_without_det_gpuaccel.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        }
+    }
+    for pipeline in cfg.test_pipeline[1:]:
+        if pipeline['type'] == 'NormalizeTensor':
+            img_metas['img_norm_cfg'] = {
+                'mean': np.array(pipeline['mean']) * 255.0,
+                'std': np.array(pipeline['std']) * 255.0
+            }
+            break
+    else:
```
[the remark this answers is not in the document]
Comment (liqikai9) [reply demo/top_down_video_demo_full_frame_without_det_gpuaccel.py:90]: I see. Here the `if` and `else` did not match in indentation, please fix it.

[the remark this answers is not in the document]
Comment (chenxinfeng4) [reply demo/top_down_video_demo_full_frame_without_det_gpuaccel.py:90]: The `else` is matched to the `for` loop, not the `if`.

File: demo/top_down_video_demo_full_frame_without_det_gpuaccel.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (liqikai9) [reply demo/top_down_video_demo_full_frame_without_det_gpuaccel.py]: Yes, this is modified in https://github.com/open-mmlab/mmpose/pull/1311, which has been merged in the master branch. Please rebase your code into the latest version of the `master` branch.

[the remark this answers is not in the document]
Comment (chenxinfeng4) [reply demo/top_down_video_demo_full_frame_without_det_gpuaccel.py]: I've aggregated the `box2cs` function in this file.

**Comment by chenxinfeng4** [comment]
Why the PR is blocked? What should I do next?

**Comment by liqikai9** [comment]
@chenxinfeng4 I found that the `ffmpegcv` is developed by you: https://github.com/chenxinfeng4/ffmpegcv. This is awesome! 
Further, I have tried the argument `--nvdecode` you add, using this command:

```
python demo/top_down_video_demo_full_frame_without_det_gpuaccel.py     configs/animal/2d_kpt_sview_rgb_img/topdown_heatmap/fly/res152_fly_192x192.py     checkpoints/res152_fly_192x192-fcafbd5a_20210407.pth     --video-path https://user-images.githubusercontent.com/87690686/165095600-f68e0d42-830d-4c22-8940-c90c9f3bb817.mp4     --show --nvdecode
```

And I got this error:


error2

Could you help check the above issue?

**Comment by chenxinfeng4** [comment]
> @chenxinfeng4 I found that the `ffmpegcv` is developed by you: https://github.com/chenxinfeng4/ffmpegcv. This is awesome! Further, I have tried the argument `--nvdecode` you add, using this command:
> 
> ```
> python demo/top_down_video_demo_full_frame_without_det_gpuaccel.py     configs/animal/2d_kpt_sview_rgb_img/topdown_heatmap/fly/res152_fly_192x192.py     checkpoints/res152_fly_192x192-fcafbd5a_20210407.pth     --video-path https://user-images.githubusercontent.com/87690686/165095600-f68e0d42-830d-4c22-8940-c90c9f3bb817.mp4     --show --nvdecode
> ```
> 
> And I got this error:
> 
> error2
> 
> Could you help check the above issue?

Please update the ffmpegcv to the latest version. I've fix some bugs of this package.

**Comment by liqikai9** [comment]
I have already used the `ffmpegcv-0.2.2` and still got the above two errors.

**Comment by chenxinfeng4** [comment]
Firstly,  the ffmpeg-writer works well in my computer (pip install ffmpegcv, version 0.2.2):
```
python demo/top_down_video_demo_full_frame_without_det_gpuaccel.py     configs/animal/2d_kpt_sview_rgb_img/topdown_heatmap/fly/res152_fly_192x192.py     checkpoints/res152_fly_192x192-fcafbd5a_20210407.pth     --video-path https://user-images.githubusercontent.com/87690686/165095600-f68e0d42-830d-4c22-8940-c90c9f3bb817.mp4     --out-video-root vis_results
```
```
pip list | grep ffmpegcv
ffmpegcv                          0.2.2
```
And got the right output video.
image

Secondly, `--nvdecode` needs the follow situation
- An proper NVIDIA GPU card. Not all NVIDIA GPU has nvidia-decoder.
- The `ffmpeg` be compiled with `--enable-nvenc`.  Most linux version didn't get such the compiled.
- The video should be 'h264' or 'h265' format. 

Your video is `mpeg4` indeed, which has no nvidia-decoder for that format.

```
ffprobe https://user-images.githubusercontent.com/87690686/165095600-f68e0d42-830d-4c22-8940-c90c9f3bb817.mp4
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from 'https://user-images.githubusercontent.com/87690686/165095600-f68e0d42-830d-4c22-8940-c90c9f3bb817.mp4':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2mp41
    encoder         : Lavf58.76.100
  Duration: 00:00:04.00, start: 0.000000, bitrate: 334 kb/s
  Stream #0:0[0x1](und): Video: mpeg4 (Simple Profile) (mp4v / 0x7634706D), yuv420p, 192x192 [SAR 1:1 DAR 1:1], 331 kb/s, 30 fps, 30 tbr, 15360 tbn (default)
    Metadata:
      handler_name    : VideoHandler
      vendor_id       : [0][0][0][0]
```

**Comment by chenxinfeng4** [comment]
Can you run this code to to test the ffmpegcv-writer? It's a CPU version, and should work well in any platform. It worked well in my computer.
```python
import ffmpegcv
videoin = 'https://user-images.githubusercontent.com/87690686/165095600-f68e0d42-830d-4c22-8940-c90c9f3bb817.mp4'
videoout = 'out.mp4'
VideoCapture = ffmpegcv.VideoCapture(videoin)
VideoWriter = ffmpegcv.VideoWriter(videoout, fps=VideoCapture.fps)

for frame in VideoCapture:
    VideoWriter.write(frame)

VideoWriter.release()
VideoCapture.release()
```

**Comment by chenxinfeng4** [comment]
Sorry, I correct my words:
>Your video is mpeg4 indeed, which has no nvidia-decoder for that format.

image

The `--nvdecode` for 165095600-f68e0d42-830d-4c22-8940-c90c9f3bb817.mp4 also works in my computer.
```
python demo/top_down_video_demo_full_frame_without_det_gpuaccel.py  configs/animal/2d_kpt_sview_rgb_img/topdown_heatmap/fly/res152_fly_192x192.py     checkpoints/res152_fly_192x192-fcafbd5a_20210407.pth     --video-path https://user-images.githubusercontent.com/87690686/165095600-f68e0d42-830d-4c22-8940-c90c9f3bb817.mp4   --nvdecode  --out-video-root vis_results
```

I guess you have something wrong with your ffmpeg installation.

**Comment by liqikai9** [comment]
@chenxinfeng4 I can not run this script in my computer yet. 
The error message is as follows:

```
(mmcv1317pt110cu102py38) PJLAB\liqikai@shai14001074l:~/open_mmlab/mmpose$ python vis_results/test_ffmpegcv.py 
Traceback (most recent call last):
  File "vis_results/test_ffmpegcv.py", line 8, in <module>
    VideoWriter.write(frame)
  File "/home/PJLAB/liqikai/anaconda3/envs/mmcv1317pt110cu102py38/lib/python3.8/site-packages/ffmpegcv/ffmpeg_writer.py", line 50, in write
    self.process.stdin.write(img)
BrokenPipeError: [Errno 32] Broken pipe

```

> Can you run this code to to test the ffmpegcv-writer? It's a CPU version, and should work well in any platform. It worked well in my computer.
> 
> ```python
> import ffmpegcv
> videoin = 'https://user-images.githubusercontent.com/87690686/165095600-f68e0d42-830d-4c22-8940-c90c9f3bb817.mp4'
> videoout = 'out.mp4'
> VideoCapture = ffmpegcv.VideoCapture(videoin)
> VideoWriter = ffmpegcv.VideoWriter(videoout, fps=VideoCapture.fps)
> 
> for frame in VideoCapture:
>     VideoWriter.write(frame)
> 
> VideoWriter.release()
> VideoCapture.release()
> ```

So, to run this script, the `ffmpeg` is a necessary prerequisite?

**Comment by chenxinfeng4** [comment]
> ```python
> import ffmpegcv
> videoin = 'https://user-images.githubusercontent.com/87690686/165095600-f68e0d42-830d-4c22-8940-c90c9f3bb817.mp4'
> videoout = 'out.mp4'
> VideoCapture = ffmpegcv.VideoCapture(videoin)
> VideoWriter = ffmpegcv.VideoWriter(videoout, fps=VideoCapture.fps)
> 
> for frame in VideoCapture:
>     VideoWriter.write(frame)
> 
> VideoWriter.release()
> VideoCapture.release()
> ```

The ffmpeg is a prerequisite. I think there may be wrong in your ffmpeg-encoders. Would you like to check the following code works for you?
```python
import cv2
import ffmpegcv
videoin = 'https://user-images.githubusercontent.com/87690686/165095600-f68e0d42-830d-4c22-8940-c90c9f3bb817.mp4'
videoout = 'out.mp4'
VideoCapture = ffmpegcv.VideoCapture(videoin)
codec = cv2.VideoWriter_fourcc(*'mp4v')
VideoWriter  = cv2.VideoWriter(videoout, codec, VideoCapture.fps, VideoCapture.size)

for frame in VideoCapture:
    VideoWriter.write(frame)

VideoWriter.release()
VideoCapture.release()
```

```python
import ffmpegcv
videoin = 'https://user-images.githubusercontent.com/87690686/165095600-f68e0d42-830d-4c22-8940-c90c9f3bb817.mp4'
videoout = 'out2.mp4'
VideoCapture = ffmpegcv.VideoCapture(videoin)
VideoWriter = ffmpegcv.VideoWriter(videoout, codec='mpeg4', fps=VideoCapture.fps) #or codec='mjpeg'

for frame in VideoCapture:
    VideoWriter.write(frame)

VideoWriter.release()
VideoCapture.release()
```

```bash
:~$ ffmpeg -encoders | grep 264
 V....D libx264              libx264 H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10 (codec h264)    <==Default encoder for ffmpegcv
 V....D libx264rgb           libx264 H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10 RGB (codec h264)
 V....D h264_nvenc           NVIDIA NVENC H.264 encoder (codec h264)        <==Default encoder for ffmpegcv NVIDIA
 V..... h264_v4l2m2m         V4L2 mem2mem H.264 encoder wrapper (codec h264)
 V....D h264_vaapi           H.264/AVC (VAAPI) (codec h264)
```

**Comment by liqikai9** [comment]
The above two python code works on my computer. And here is my ffmpeg:

```
(mmcv1317pt110cu102py38) PJLAB\liqikai@shai14001074l:~/open_mmlab/mmpose$ ffmpeg -encoders | grep 264
ffmpeg version 4.3 Copyright (c) 2000-2020 the FFmpeg developers
  built with gcc 7.3.0 (crosstool-NG 1.23.0.449-a04d0)
  configuration: --prefix=/opt/conda/conda-bld/ffmpeg_1597178665428/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh --cc=/opt/conda/conda-bld/ffmpeg_1597178665428/_build_env/bin/x86_64-conda_cos6-linux-gnu-cc --disable-doc --disable-openssl --enable-avresample --enable-gnutls --enable-hardcoded-tables --enable-libfreetype --enable-libopenh264 --enable-pic --enable-pthreads --enable-shared --disable-static --enable-version3 --enable-zlib --enable-libmp3lame
  libavutil      56. 51.100 / 56. 51.100
  libavcodec     58. 91.100 / 58. 91.100
  libavformat    58. 45.100 / 58. 45.100
  libavdevice    58. 10.100 / 58. 10.100
  libavfilter     7. 85.100 /  7. 85.100
  libavresample   4.  0.  0 /  4.  0.  0
  libswscale      5.  7.100 /  5.  7.100
  libswresample   3.  7.100 /  3.  7.100
 V..... libopenh264          OpenH264 H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10 (codec h264)

```

**Comment by liqikai9** [comment]
BTW, if ffmpeg is a prerequisite, please also note it in the code to guide the users. Thanks~

**Comment by chenxinfeng4** [comment]
> 

Ohh, I see. The ffmpeg in your environment is **imcomplete**. It doesn't support GPU, and lacks many video encoders and decoders. 

Here is my ffmpeg.
To simplify this issue, I will specify the ffmpeg-writer to the general 'mpeg4' codec in this PR.
```bash
$ /bin/ffmpeg
ffmpeg version 4.2.4-1ubuntu0.1 Copyright (c) 2000-2020 the FFmpeg developers
  built with gcc 9 (Ubuntu 9.3.0-10ubuntu2)
  configuration: --prefix=/usr --extra-version=1ubuntu0.1 --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu \
--incdir=/usr/include/x86_64-linux-gnu --arch=amd64 --enable-gpl --disable-stripping --enable-avresample \
--disable-filter=resample --enable-avisynth --enable-gnutls --enable-ladspa --enable-libaom --enable-libass \
--enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libcodec2 --enable-libflite \
--enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libjack \
--enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus \
--enable-libpulse --enable-librsvg --enable-librubberband --enable-libshine --enable-libsnappy \
--enable-libsoxr --enable-libspeex --enable-libssh --enable-libtheora --enable-libtwolame \
--enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp \
--enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzmq --enable-libzvbi \
--enable-lv2 --enable-omx --enable-openal --enable-opencl --enable-opengl --enable-sdl2 \
--enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-nvenc --enable-chromaprint \
--enable-frei0r --enable-libx264 --enable-shared
```

**Comment by liqikai9** [comment]
@chenxinfeng4 I have rebased the PR into the latest master branch and added descriptions in the related docs. 

This PR looks good to me now. Please check it. If you do not have other issues, I will approve this PR.

**Comment by chenxinfeng4** [comment]
It looks good to me.

**Review by liqikai9 — APPROVED** [review]

**Pull request merged**